### PR TITLE
[Experimental] Pluggable package resolution for fsi and VF# editor

### PR DESCRIPTION
**Pull request by amazingant**
This is a hefty change to @forki's PR at #2483, and basically just rearranges some of his code. In his new file `ReferenceLoading.PaketHandler.fs`, I added a short interface named `IPragmaProvider`, and converted the WIP paket code into a new typedef that implements this interface. Expectation is that the `IPragmaProvider` interface and perhaps a new attribute would live somewhere that third-party libraries can reference and implement the interface, much like type providers do now. But of course I have no clue how to do that, or where to look for examples.

This PR is up-to-date with #2483 as of about a half hour ago; @forki works too fast for me to keep up 😉 

General disclaimer:

no idea what im doing

### CODE REVIEW

**Comment by forki** [comment]
Ok what's the difference to mine? And could you please send it against my fork (after you signed the CLA) - we integrate over my fork at the moment.

**Comment by forki** [comment]
Tomorrow I fix the Paket issue

**Comment by amazingant** [comment]
@forki will move the PR shortly; just got home.

Main difference is that instead of just building the paket code into `FSharp.Compiler`, my commit creates an `IPragmaProvider` interface and builds a new type that implements it. The new `PaketPragmaHandler` type acts as a sort of plugin and adds a new `#paket` pragma. With my commit, running this through fsi works:

```FSharp
#paket "nuget NUnit"
typeof<NUnit.Framework.ActionTargets>.Assembly.Location
```

and outputs the file path to the nunit library where your code stashes it. This really is intended to build off the comment I made this morning [over here](https://github.com/fsharp/fslang-suggestions/issues/542#issuecomment-282717209), suggesting to make this extensible somehow.

**Comment by smoothdeveloper** [comment]
@amazingant wow!

I agree with @forki, making a PR on his fork is the way to go.

@forki for scriptcs integration, for now I had this disclaimer:
```fsharp
// NOTE: this file is used by other parties integrating paket reference loading in scripting environments.
// Do not add any reference to F# codebase other than FSharp.Core.
// This file might end up in paket repository instead of F#.
```

I like that in his PR, @amazingant seem to have changed the mutable state that we had over CompileOps and encapsulated this in the PaketPragma implementation (much smaller scope).

I think we will need to split the file containing the pragma interface / paket implementation as they are closer to language / compiler concern than just the paket execution logic.

I think ultimately the paket execution logic should land in Paket repository and we have just the wrapping as pragma extensions implemented + integration points in CompileOps/fsi.

I haven't looked at the other changes, we should direct that to @forki's fork.

@amazingant if you can, try to put the interface/paket pragma implementation in a separate file as I doubt this will be easily consumable outside F# compiler.

**Comment by amazingant** [comment]
@smoothdeveloper will do; if @forki isn't making any more huge changes tonight, is it okay if I continue on this when I get back to work tomorrow morning? Waiting on a response from my supervisor as to whether or not I'm allowed to sign the CLA, and not eager to spend another day giving myself headaches without getting his answer. 😬

**Comment by forki** [comment]
I'm done for today.

Am 27.02.2017 22:16 schrieb "Anthony Perez" <notifications@github.com>:

**Comment by forki** [comment]
when you send the PR to my fork, please make sure you take a look at the test samples at `tests\fsharpqa\Source\InteractiveSession\Paket` thx

File: src/fsharp/ReferenceLoading.PaketHandler.fs [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    /// baseDir/.paket/load/frameworkDir/scriptName 
+    let GetPaketLoadScriptLocation baseDir optionalFrameworkDir scriptName =
+      let paketLoadFolder = System.IO.Path.Combine(Internals.PM_DIR,"load")
+      getLoadScript
+        baseDir
+        (match optionalFrameworkDir with | None -> paketLoadFolder | Some frameworkDir -> System.IO.Path.Combine(paketLoadFolder, frameworkDir))
+        scriptName
+
+
+module GenericHandler =
```
Comment (dsyme) [inline src/fsharp/ReferenceLoading.PaketHandler.fs]: This should probably be in CompilerOps.fs/fsi

File: src/fsharp/ReferenceLoading.PaketHandler.fs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    type IPragmaProvider =
+        inherit System.IDisposable
+        abstract PragmaName : string
+        abstract ProcessArg : string -> unit
+        abstract HasPendingChanges : bool
+        abstract ChangesToScript : string -> (string * string list) option
+
+    type PaketPragmaHandler() =
```
Comment (dsyme) [inline src/fsharp/ReferenceLoading.PaketHandler.fs]: With this implementation in the Paket-specific file

**Comment by dsyme** [comment]
@amazingant With regards to ``#paket`` here https://github.com/Microsoft/visualfsharp/pull/2504/files#diff-52a3b381c6bc6da5211d19ccdec6484dR1987

I know this is just an experiment, so anything is ok for now, but my intuition is that I'd rather not enable ``#paket`` - I'd prefer just ``#r "paket: ..."``

**Comment by amazingant** [comment]
@dsyme re `#paket` vs `#r "paket: ..."`, I was going off my suggestion (originally @Rickasaurus's) on fsharp/fslang-suggestions#542 to provide a way to plug in new pragma types, but I can revert to `#r "paket: ..."` before I open a PR to @forki's fork.

Would it be acceptable for fsi/fsc to detect types that implement a new prefix as it does with `TypeProviderAssemblyAttribute` and `ITypeProvider`? Again, I'm thinking of this as a pluggable interface where e.g. I could make my own `#r "amazingant: ..."` prefix by just implementing the interface, rather than limiting it to whichever prefixes get merged into `master`.

**Comment by forki** [comment]
I think the current idea is:

     #r "amazingant: ..."  // should defer calls to amazingant.dll (or other naming pattern)

and

    #amazingant: // not accepted

So it's just about not introducing new syntax at this point in time while still getting rid of anything that directly binfs to paket

**Comment by forki** [comment]
I actually think a sample with a `#r amazingant: ` would be valueable, since it would show that this is not boundled.

**Comment by amazingant** [comment]
@forki sounds good; I originally came here to implement it as `#amazingant`, but working it up as `#r "amazing: ..."` still allows me to ~abuse~ make use of the feature for just as ~nefarious~ useful purposes. 😄

Still waiting for supervisor to get in and see if I can get an answer on the CLA, but will start on code changes while I wait.

**Comment by smoothdeveloper** [comment]
@amazingant / @forki on that line I'm thinking:

* it would be good PoC to have the existing directives implement the infrastructure @amazingant created, it seems cleaner down the road to have that, and would show that the technical solution is sound, although we are not yet eager to add all kinds of pragma to the compiler (there will be clash in the eco system for sure if we do that)
* I'm also in favor of reusing `#r` but can we consider `#r paket:"..."` at some point, I think it would give us better direction and fit nice with idea of having more extensions of existing pragmas, not making it all string based

AFAIK, in haskell, the top level pragmas are fixed / conservative, but what comes after has more variation / things added on shorter notice, this seem reasonable approach for F# as well.

**Comment by dsyme** [comment]
> Would it be acceptable for fsi/fsc to detect types that implement a new prefix as it does with TypeProviderAssemblyAttribute and ITypeProvider? Again, I'm thinking of this as a pluggable interface where e.g. I could make my own #r "amazingant: ..." prefix by just implementing the interface, rather than limiting it to whichever prefixes get merged into master.

I'm torn about whether this needs to be a new binary interface, or just a protocol to invoke a ``paket.fsi.exe`` and ``dotnet.fsi.exe`` and  ``amazingant.fsi.exe``.  Perhaps an attributed binary interface is easy enough, as long as the components will load into both a .NET Framework and .NET Core ``fsi.exe``.  I think it doesn't really matter that much as long as it's robust, performant and gives good diagnostics.

File: src/fsharp/ReferenceLoading.PaketHandler.fs (hunk 1 above)
Comment (smoothdeveloper) [inline src/fsharp/ReferenceLoading.PaketHandler.fs]: @dsyme "putting more code in CompileOps/fsi considered harmful" :)

For someone trying to contribute, dealing with those huge compact files is really daunting, if we use small files (and some folder structure) I think the codebase would tend to become more manageable.

In my day job, I also deal with large files at times, when I see a change just adding some line of code, which could instead have been put in a separate component with only small insertion point fitting better the level at integration point, I feel the change was done in haste without considering that option.

I'm not advocating the C#/Java approach here though, maybe we can consider a PragmaHandling.fs module with the code you are pointing.

File: src/fsharp/ReferenceLoading.PaketHandler.fs (hunk 2 above)
Comment (smoothdeveloper) [inline src/fsharp/ReferenceLoading.PaketHandler.fs]: Yes, in the one for F# compiler (which is not src/fsharp/ReferenceLoading.PaketHandler.fs but the other one I added with few small members).

**Comment by forki** [comment]
We can always improve on syntax on the language design issue. This one here
should focus on the technical things. Extracting paket specifics, open up
api, proper error  testing,...

Am 28.02.2017 13:49 schrieb "Gauthier Segay" <notifications@github.com>:

**Comment by amazingant** [comment]
@dsyme invoking `*.fsi.exe` seems fragile to me, but it wouldn't be any harder to do I suppose. If the `*.fsi.exe` assemblies wanted to avoid providing duplicate information, they'd need to somehow keep track of the inputs they had been previously given, likely on-disk somewhere. So e.g. if I ran `#r "paket: NUnit"` and did some stuff, then ran `#r "paket: DotNetZip"`, could `paket.fsi.exe` at some point understand that we do not need to find `NUnit` again since it has already been loaded? Even if `paket.fsi.exe` does not support that at first, I think it would be easier to add that kind of behavior to it if the data remained in-memory by loading an interface.

File: src/fsharp/ReferenceLoading.PaketHandler.fs (hunk 1 above)
Comment (forki) [inline src/fsharp/ReferenceLoading.PaketHandler.fs]: please let us stick to current overall principles in the compiler. we should not introduce new battlefields for this PR

**Comment by forki** [comment]
hey @amazingant what about the CLA? ;-)

meanwhile i progressed a bit with my PR. There is a PackageManager.Integration.fs that already goes into the right direction. I think you should look at that and creates your hooks over there. thanks

**Comment by amazingant** [comment]
@forki supervisor got back to me yesterday morning and basically said he has no problem with me signing it, but that he isn't actually allowed to say so. So the request has gone up the corporate ladder. 😐

Hopefully you're not completely done by the time I get an answer, but knowing how you work... 😉

**Comment by forki** [comment]
I think there is a option for when you do it on your free time that does
not involve manager approval. And no faxing

Am 01.03.2017 16:14 schrieb "Anthony Perez" <notifications@github.com>:

**Comment by amazingant** [comment]
@forki My employer requires an intellectual property contract at the time of employment, so while I could probably get away with it, I'm not excited by the idea of fighting my employer over this, now that they know about it. 😬

Of course if you know a lawyer willing to look at everything for free and tell me it's safe to just do the work at home, I'd happily accept. 😄

**Comment by forki** [comment]
lol. you are in the US of A I guess ;-)

2017-03-01 17:41 GMT+01:00 Anthony Perez <notifications@github.com>:

**Comment by amazingant** [comment]
@forki sadly, yes

**Comment by amazingant** [comment]
Latest news on the CLA is a meeting with someone from our legal department next week. Closing this issue as @forki is already making good progress on making this pluggable, hopefully will have a signed CLA before I open any other PRs.

**Comment by KevinRansom** [comment]
@amazingant regarding your CLA, did you get it sorted out?

Kevin

**Comment by amazingant** [comment]
@KevinRansom Yes and no. The meeting was a bit of a waste of time, but the short answer is that I am allowed to sign the CLA, but must get VP-level approval on a case-by-case basis for each contribution. Not the end of the world, but @forki has already worked up the gist of what I was trying to accomplish here. :)

In the future, if there is a contribution I can make that nobody else can, I'll take the time to climb the ladder to my VP, but it seemed like an endless black hole of meetings for this small patch.

**Comment by KevinRansom** [comment]
Oh my goodness ... I'm sorry to hear that, we should encourage your VP to become a regular contributor too.  We need all of the help we can get :-)

Kevin

**Pull request closed without merging**